pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,260 @@
+			// Create a reconcile.Request for the router named in the RouteIngress.
+			request := reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      routerName,
+					Namespace: r.Namespace,
+				},
+			}
+			requests = append(requests, request)
+		}
+	}
+
+	// Map the currentRouteIngresses to Route's NamespacedName.
+	r.routeToIngresses[routeNamespacedName] = currentRouteIngresses
+
+	return requests
+}
+
+// reconciler handles the actual ingresscontroller reconciliation logic in response to events.
+type reconciler struct {
+	cache     cache.Cache
+	client    client.Client
+	Namespace string
+	// routeToIngresses stores the Ingress Controllers that have admitted a given route.
+	routeToIngresses map[types.NamespacedName]sets.String
+}
+
+// Reconcile expects request to refer to an Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.cache.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get Ingress Controller %q: %v", request, err)
+	}
+
+	// If the ingresscontroller is marked to be deleted, then delete the corresponding RouteMetricsControllerRoutesPerShard metric label.
+	if ingressController.DeletionTimestamp != nil {
+		// If the routemetricscontroller finalizer is not removed, then delete the corresponding RouteMetricsControllerRoutesPerShard metric label
+		// and remove the finalizer.
+		if controllerutil.ContainsFinalizer(ingressController, manifests.RouteMetricsControllerFinalizer) {
+			// Delete the Shard label corresponding to the Ingress Controller from the RouteMetricsControllerRoutesPerShard metric.
+			DeleteRouteMetricsControllerRoutesPerShardMetric(request.Name)
+			log.Info("RoutesPerShard metric label corresponding to the Ingress Controller is successfully deleted", "ingresscontroller", ingressController)
+
+			// Remove the routemetricscontroller finalizer.
+			updated := ingressController.DeepCopy()
+			controllerutil.RemoveFinalizer(updated, manifests.RouteMetricsControllerFinalizer)
+			err := r.client.Update(ctx, updated)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// Before doing anything at all with the controller, ensure it has a finalizer
+	// so we can clean up later.
+	if !controllerutil.ContainsFinalizer(ingressController, manifests.RouteMetricsControllerFinalizer) {
```

Comment:
I'm trying to understand why we'd need to check for this in the Reconcile loop?

Reply:
Moved the deletion of metrics to `ensureIngressDeleted`.